### ISSUE TITLE
Is it possible to manually update the cache with a *nested* object from a subscription update?

### ISSUE THREAD

**Issue by thormuller** [AlecAivazis/houdini#155]
On Alec's recommendation, I previously setup a subscription that updates when new records are created, the result were then added manually to the cache. Example: https://github.com/AlecAivazis/houdini/issues/94#issuecomment-859989593.  Like in this example the query and its results were flat (single level) and the manual prepend to the cache worked well. 

I have since updated my schema with nested queries. Something like this:

```
const { data } = subscription(
graphql`
	subscription NewRecipe($id: String!) {
		queryNewRecipe(id: $id) {
			id
			versions(first: 1) {
				title
				id
			}
			sourceRecipe {
				id
				domain
			}
		}
	}
`,
{
	id: $session.user.id
}
);

$: {
    cache.list('Active_Recipes').prepend(
	    {
		    id: { type: 'ID', keyRaw: 'id' },
		    versions: [
			    {
				    title: { type: 'String', keyRaw: 'title' },
				    id: { type: 'ID', keyRaw: 'id' },
			    }
		    ],
		    sourceRecipe: {
			    id: { type: 'ID', keyRaw: 'id' },
			    domain: { type: 'String', keyraw: 'domain' }
		    }
	    },
	    $data.queryNewRecipe
    );
}
```
Unfortunately, now the manual prepend to the cache is failing with `TypeError: key is not iterable`. When I inspect cache.js it looks like the code doesn't really handle nested schemas like this, as the keys and values are showing as `undefined` inside the `insertSubscribers` method.

**Assuming I understand the limitation correctly, how would you now recommend I manually update the cache with the incoming data to given its structure?**

The full error:
```
Uncaught (in promise) TypeError: key is not iterable
    at Cache.evaluateKey (cache.js?t=1629053326512:713)
    at Cache.insertSubscribers (cache.js?t=1629053326512:665)
    at ListHandler.addToList (list.js:107)
    at ListHandler.prepend (list.js:33)
    at Object.$$self.$$.update (Subscriptions.svelte:26)
    at update (index.mjs?v=8f41bf68:1046)
    at flush (index.mjs?v=8f41bf68:1018)
```

**Comment by AlecAivazis** [comment]
Hey @thormuller - I’m currently traveling but will have more stable internet in a few days to look into this. The cache should support your use case well just have to structure the update a little differently to store the associations.

**Comment by thormuller** [comment]
Awesome, thanks. Enjoy your travels!

**Comment by AlecAivazis** [comment]
Hey @thormuller  - looking at your example, i don't see the same  unique constraint on the updater. Given this, I think it would be a lot easier to use the `Active_Recipes_insert` fragment. Either way, I think the issue you are running into is with the selection object you are passing in. I think you should be passing something like:

```
{
	id: { type: 'ID', keyRaw: 'id' },
	versions:  {
		title: { type: 'String', keyRaw: 'title' },
		id: { type: 'ID', keyRaw: 'id' },
	},
	sourceRecipe: {
		id: { type: 'ID', keyRaw: 'id' },
		domain: { type: 'String', keyRaw: 'domain' }
	}
}
```

**Comment by thormuller** [comment]
Thanks, this makes sense. However, I'm unable to test it now because in the course of all this I I updated Svelte Kit and now when I run dev I get the following Houdini error:

`environment.ts:1 Uncaught (in promise) SyntaxError: The requested module '/$houdini/index.js' does not provide an export named 'SubscriptionHandler'`

Is there an obvious debugging path I should know about here?

**Comment by AlecAivazis** [comment]
Did you update your houdini version as well? You could always try to delte the `.svelte-kit` directory in your project's root

**Comment by thormuller** [comment]
Houdini was already updated to the latest version and working well prior to the svelte-kit update. Deleting `.svelte-kit` didn't solve it, alas. I'll keep poking at it and see what I find.

**Comment by thormuller** [comment]
Incidentally, I'm seeing the same error appear when running the Houdini **example app** after I update the Sveltekit version to `1.0.0-next.120` or later, which makes me think it may be something in my dev environment. Note, when I run the example with the default sveltekit version in the package (`1.0.0-next.107`) I get the following CommonJS error:

```
 import svelte from '@sveltejs/vite-plugin-svelte';
 SyntaxError: The requested module '@sveltejs/vite-plugin-svelte' does not provide an export named 'default'
```

**Comment by AlecAivazis** [comment]
I believe @pixelmund is using `141` so I dont think there is some kind of incompatibility. I'll try to find the time to update the example app this week

**Comment by thormuller** [comment]
I think I found the issue. 

The documentation for subscriptions specifies that the following imports should be amended to the top of `environment.ts`:
```
import { SubscriptionClient } from 'subscriptions-transport-ws'
import { browser } from '$app/env'
```

However, the example app uses a different construction, which I'd replicated in my code:
```
import { Environment, **SubscriptionHandler** } from '$houdini'
import { SubscriptionClient } from 'subscriptions-transport-ws'
import { browser } from '$app/env'
```
Removing the named import for SubscriptionHandler fixes the issue, and things work fine. The good news is the documentation is correct; the bad news is the example is misleading and (at least in my local setup) doesn't work. I'm not sure why it used to work and doesn't now.

**Comment by AlecAivazis** [comment]
Huh, okay thanks for digging into this. I'll try to figure out what's going on in the example. If you hover over the type import does your ide find it? Like you said, it seems weird it would randomly break

**Comment by thormuller** [comment]
Yes, subscriptions are working well now, and the IDE does see the SubscriptionHandler for that named import. 

In terms of dev environment changes I updated svelte kit, npm, and a few dependencies directly preceding when I started seeing this issue. Here are my versions in case it's useful (node v16.4.0):

├── @apollo/client@3.4.8
├── @sveltejs/adapter-node@1.0.0-next.42
├── @sveltejs/kit@1.0.0-next.153
├── autoprefixer@10.3.1
├── cookie@0.4.1
├── graphql@15.5.0
├── houdini-preprocess@0.10.4
├── houdini@0.10.4
├── path@0.12.7
├── postcss-load-config@3.1.0
├── postcss@8.3.6
├── subscriptions-transport-ws@0.9.18
├── svelte-preprocess@4.7.4
├── svelte@3.42.1
├── tslib@2.3.1
├── typescript@4.3.5
└── uuid@8.3.2

**Comment by AlecAivazis** [comment]
fantastic this is super helpful. I'm going to go ahead and close the ticket since your issue has been resolved. Thanks again!

**Comment by AlecAivazis** [comment]
I'd like to understand a little better what's causing you to use a manual update instead of the fragment. Are you still trying to ensure unique-ness in a connection?

**Comment by thormuller** [comment]
To jog your memory: https://github.com/AlecAivazis/houdini/issues/94#issuecomment-859069148

TLDR; This is the consequence of a limitation on the subscription result which doesn't play well with the insert method (it leads to duplicate results in this context).

**Comment by AlecAivazis** [comment]
Ah okay, i remembered your situation i just didn't see the same check in your recent example. I'm still trying to figure out the right solution for your use case

**Comment by thormuller** [comment]
If the graphql server/db I'm using should just support subscriptions on Add/Update/Delete this would be a non-issue from Houdini's perspective. That seems like the "right" solution to me :)

**Comment by thormuller** [comment]
@AlecAivazis I decided to try to get the Insert method working, since the current setup is too klugey (I'm okay abandoning some desired functionality in the short term for more compliant code). The good news is that new items are added correctly to the list. **However**, subscription updates to freshly inserted items are not updating the displayed items in the list. Subscription updates ARE updating the items that are pre-loaded. AND the cache is showing the records reflecting the subscription updates. It's just that the _displayed_ list items are not updated immediately following insert.

I tried to compare this behavior to the Example app insert functionality, but alas I'm having trouble launching it (the example app is not successfully connecting to the Apollo server api despite the web app working fine-showing a 404). In the event that you have the example app working, it would be great if you can confirm that this does work on your end, before I continue to dig into my particulars.

**Comment by AlecAivazis** [comment]
Huh, that's definitely strange. I just verified that it's working on my end. If you look at the record's field's subscriber object do you see your query's set registered?

**Comment by thormuller** [comment]
Looks like it the subscriber is _not_ being set on the freshly inserted items. Here's what that record looks like: ![image](https://user-images.githubusercontent.com/395315/130315624-6605f68d-3599-4e4d-b868-5429f329f7cb.png)

Versus an item that is part of the initial query list :
![image](https://user-images.githubusercontent.com/395315/130315680-590b5384-1be5-46ca-aadd-b2b09e9eeba3.png)

In case it's useful, here's what the queries look like. Preliminary:
```
	const { data } = query<AllMyRecipes>(graphql`
		query AllMyRecipes($id: String!) {
			activeRecipes: queryMyRecipes(id: $id) @list(name: "Active_Recipes") {
				id
				...Card_recipe
			}
		}
	`);
```

The fragment:
```
	const data = fragment(
		graphql`
			fragment Card_recipe on Recipe {
				versions(first: 1) {
					title
					id
					image
				}
				sourceRecipe {
					id
					domain
				}
				id
			}
		`,
		recipe
	);
```

And the subscription:
```
	subscription(
		graphql`
			subscription RecipeUpdate($id: ID!) {
				getRecipe(id: $id) {
					id
					...Card_recipe
				}
			}
		`,
		{
			id: recipe.id
		}
	);
```

**Comment by thormuller** [comment]
I just notice that the example app has a different pattern than what is shown in the documentation. In the documentation the insert is in the mutation, while in the example the insert is in the subscription.

I can try implementing it the way it is in the example since you've confirmed that works.

**Comment by AlecAivazis** [comment]
The subscription and mutation logic both rely on the same function that writes data which is responsible for detecting new records that need subscriptions. Would you be able to publish a repo that shows the issue so i can dig a little deeper?

**Comment by thormuller** [comment]
Hi Alec,
I sent you an email with access and notes to check it out. Thanks!
